File: frontend/src/types/challenge.ts
export interface CategoryDto {
  id: number;
  name: string;
  description: string | null;
  iconUrl: string | null;
}

export interface DifficultyDto {
  id: number;
  name: string;
  color: string | null;
  xpMultiplier: number;
}

export interface ChallengeListItemDto {
  id: number;
  title: string;
  category: string;
  difficulty: string;
  xpReward: number;
  coinReward: number;
  isPublished: boolean;
  requiredLevel: number;
  isLocked: boolean;
  tags: string | null;
  isBattleOnly: boolean;
}

export interface TestCaseDto {
  id: number;
  input: string;
  expectedOutput: string;
  orderIndex: number;
}

export interface HiddenTestCaseDto {
  id: number;
  input: string;
  expectedOutput: string;
  orderIndex: number;
  weight: number;
}

export interface ChallengeStatsDto {
  solverCount: number;
  totalSubmissions: number;
  acceptanceRate: number;
  averageSolveTimeMs: number | null;
}

export interface ChallengeCommentDto {
  id: number;
  content: string;
  isSpoiler: boolean;
  createdAt: string;
  userId: number;
  username: string;
  avatarUrl: string | null;
  parentCommentId: number | null;
}

export interface ChallengeCommentThreadDto {
  comment: ChallengeCommentDto;
  replies: ChallengeCommentDto[];
}

export interface DailyPuzzleDto {
  challengeId: number | null;
  title: string | null;
  difficulty: string | null;
  solvedToday: boolean;
}

export interface ChallengeDetailDto {
  id: number;
  title: string;
  description: string;
  categoryId: number;
  category: string;
  difficultyId: number;
  difficulty: string;
  timeLimitMs: number;
  memoryLimitMb: number;
  xpReward: number;
  coinReward: number;
  starterCode: string;
  constraints: string | null;
  inputFormat: string | null;
  outputFormat: string | null;
  sampleInput: string | null;
  sampleOutput: string | null;
  isPublished: boolean;
  requiredLevel: number;
  testCases: TestCaseDto[];
  hiddenTestCases: HiddenTestCaseDto[] | null;
  isAlreadySolved: boolean;
  // Admin-only: raw (unresolved) English variants, populated only when the requester is an Admin.
  titleEn: string | null;
  descriptionEn: string | null;
  constraintsEn: string | null;
  inputFormatEn: string | null;
  outputFormatEn: string | null;
  starterCodeEn: string | null;
  tags: string | null;
  isBattleOnly: boolean;
}

File: backend/src/QuestCraft.Domain/Entities/Challenge.cs
using QuestCraft.Domain.Common;

namespace QuestCraft.Domain.Entities;

public class Challenge : BaseEntity
{
    public string Title { get; set; } = default!;
    public string Description { get; set; } = default!;
    public int TimeLimitMs { get; set; } = 2000;
    public int MemoryLimitMb { get; set; } = 256;
    public int XpReward { get; set; }
    public int CoinReward { get; set; }
    public string StarterCode { get; set; } = default!;
    public string? Constraints { get; set; }
    public string? InputFormat { get; set; }
    public string? OutputFormat { get; set; }
    public string? SampleInput { get; set; }
    public string? SampleOutput { get; set; }
    public bool IsPublished { get; set; }
    public int RequiredLevel { get; set; } = 1;

    // Comma-separated, e.g. "linq,collections,beginner" — simple free-text tagging rather than a
    // normalized many-to-many table, since search/filter is the only consumer and volume is small.
    public string? Tags { get; set; }

    // Battle-pool challenges are a separate set from the leveled practice list: excluded from
    // GetChallengesQuery entirely (never appear at any level) and drawn from at random when a
    // Battle room/duel is created, so no participant could have already seen or solved it.
    public bool IsBattleOnly { get; set; }

    // Daily-puzzle-pool challenges are also excluded from GetChallengesQuery. Which one is "today's"
    // puzzle is never stored — it's derived deterministically from the date (see GetDailyPuzzleQuery),
    // so this flag only marks pool membership, not the currently-active pick.
    public bool IsDailyPuzzle { get; set; }

    // English translations — nullable; falls back to the Azerbaijani field above when missing.
    public string? TitleEn { get; set; }
    public string? DescriptionEn { get; set; }
    public string? ConstraintsEn { get; set; }
    public string? InputFormatEn { get; set; }
    public string? OutputFormatEn { get; set; }
    public string? StarterCodeEn { get; set; }

    public int CategoryId { get; set; }
    public ChallengeCategory Category { get; set; } = default!;

    public int DifficultyId { get; set; }
    public ChallengeDifficulty Difficulty { get; set; } = default!;

    public ICollection<TestCase> TestCases { get; set; } = new List<TestCase>();
    public ICollection<HiddenTestCase> HiddenTestCases { get; set; } = new List<HiddenTestCase>();
    public ICollection<ChallengeSubmission> Submissions { get; set; } = new List<ChallengeSubmission>();
}
